File: python/spot_it.py
# this is code to generate "Spot It!" cards
# using the finite projective plane.
# 
# The finite projective plane of order n (where n is prime)
# contains n^2 + n + 1 points and n^2 + n + 1 lines.
# Each line goes through n + 1 points
# Each point lies on n + 1 lines
# Any two lines intersect in exactly one point
#
# Now think:
#  point = picture
#  line = card

# for more details, see the blog post:
#
#

#
# POINTS
#

def ordinary_points(n):
    """ordinary points are just pairs (x, y) where x and y
    are both between 0 and n - 1"""
    return [(x, y) for x in range(n) for y in range(n)]
    
def points_at_infinity(n):
    """infinite points are just the numbers 0 to n - 1
    (corresponding to the infinity where lines with that slope meet)
    and infinity infinity (where vertical lines meet)"""
    return range(n) + [u"∞"]

def all_points(n):
    return ordinary_points(n) + points_at_infinity(n)

def ordinary_line(m, b, n):
    """returns the ordinary line through (0, b) with slope m
    in the finite projective plan of degree n
    includes 'infinity m'"""
    return [(x, (m * x + b) % n) for x in range(n)] + [m]

def vertical_line(x, n):
    """returns the vertical line with the specified x-coordinate
    in the finite projective plane of degree n
    includes 'infinity infinity'"""
    return [(x, y) for y in range(n)] + [u"∞"]
    
def line_at_infinity(n):
    """the line at infinity just contains the points at infinity"""
    return points_at_infinity(n)

def all_lines(n):
    return ([ordinary_line(m, b, n) for m in range(n) for b in range(n)] +
            [vertical_line(x, n) for x in range(n)] +
            [line_at_infinity(n)])

import random

# get a list of animals we can use to name the cards
with open("animals.txt", "r") as f:
    animals = [line.strip() for line in f]

def make_deck(n, pics=None):
    points = all_points(n)

    # if no pics specified, start with empty list
    if pics is None: pics = []

    # if fewer pics than points, add indexes
    if len(pics) < len(points):
        pics.extend(range(len(pics), len(points)))

    # then take a random sample        
    pics = random.sample(pics, len(points))

    # create a mapping from point to pic
    mapping = { point : pic 
                for point, pic in zip(points, pics) }

    # and return the remapped cards
    return [map(mapping.get, line) for line in all_lines(n)]
    
def play_game(deck):
    # make a copy so as not to much with the original deck,
    # and then shuffle it
    deck = deck[:]
    random.shuffle(deck)

    # keep playing until fewer than 2 cards are left    
    while len(deck) >= 2:
        card1 = deck.pop()
        card2 = deck.pop()
        random.shuffle(card1)
        random.shuffle(card2)

        # find the matching element
        match, = [pic for pic in card1 if pic in card2]  

        print card1
        print card2
        
        guess = raw_input("Match? ")
        if guess == match:
            print "correct!"
        else:
            print "incorrect!"

File: haskell/SpotIt.hs
module SpotIt where

import qualified Data.Map as Map

-- | We define three types of points:
-- * an ordinary point (x, y)
-- * a point at infinity corresponding to a slope m
-- * the point at infinity corresponding to infinite slope
data Point = OrdinaryPoint Int Int  -- x, y
           | PointAtInfinity Int    -- m
           | VerticalInfinity       --
    deriving (Ord, Eq, Show)

-- | Similarly, we define three types of lines:
-- * an ordinary line with slope m and intercept b
-- * a vertical line through (x, 0)
-- * the line at infinity
data Line = OrdinaryLine Int Int    -- m, b
          | VerticalLine Int        -- x
          | LineAtInfinity          -- 
    deriving (Ord, Eq, Show)

-- | a helper function to create the list from 0 to n - 1
upTo :: Int -> [Int]
upTo n = [0 .. (n - 1)]

-- | there are n^2 ordinary points (x, y)
ordinaryPoints :: Int -> [Point]
ordinaryPoints n = [OrdinaryPoint x y | x <- upTo n, y <- upTo n]

-- | along with n + 1 points at infinity, one for each slope
infinitePoints :: Int -> [Point]
infinitePoints n = VerticalInfinity : [PointAtInfinity m | m <- upTo n]

-- | combine these to get all points
allPoints :: Int -> [Point]
allPoints n = ordinaryPoints n ++ infinitePoints n

-- | we can do the same to get all of the lines:
-- * n^2 ordinary lines,
-- * n vertical lines,
-- * 1 line at infinity
allLines :: Int -> [Line]
allLines n = ordinaryLines n ++ verticalLines n ++ [LineAtInfinity] where
    ordinaryLines n = [OrdinaryLine m b | m <- upTo n, b <- upTo n]
    verticalLines n = [VerticalLine x | x <- upTo n]

-- | given n and a Line, return the points on the line
-- three different cases depending on the type of line
pointsOnLine :: Int -> Line -> [Point]
pointsOnLine n (OrdinaryLine m b) = PointAtInfinity m : 
    [OrdinaryPoint x ((m * x + b) `mod` n) | x <- upTo n]
pointsOnLine n (VerticalLine x) = VerticalInfinity :
    [OrdinaryPoint x y | y <- upTo n]
pointsOnLine n LineAtInfinity = infinitePoints n

-- type aliases for our game constructs
type Picture = String
type Card = [Picture]
type Deck = [Card]

-- | Given n and a list of "pictures", create the deck of order n by
-- * finding all n^2 + n + 1 points
-- * creating a Map : Point -> Picture
-- * finding all n^2 + n + 1 lines 
-- * mapping each line to the list of points it contains
-- * mapping each list of points to the list of corresponding pictures
createDeck :: Int -> [Picture] -> Deck
createDeck n picNames = map (remap . pointsOnLine n) $ allLines n
    where
        encoding = Map.fromList $ zip (allPoints n) picNames
        remap = map (encoding Map.!)
                
-- For instance, to create a deck with "pictures" labeled 0 to 56, we could do
deck :: Deck
deck = createDeck 7 (map show [0..])

-- | Given two cards, return the unique picture they have in common
-- OK to use unsafe head because intersection is always one picture
picInCommon :: Card -> Card -> Picture
picInCommon card1 card2 = head commons
    where
        commons = [pic1 | pic1 <- card1, pic2 <- card2, pic1 == pic2]

-- | Given a deck, play a game
play :: Deck -> IO ()
play (card1:card2:cards) = do
    putStrLn $ show card1
    putStrLn $ show card2
    guess <- getLine
    putStrLn $ if guess == picInCommon card1 card2 then "correct!" else "wrong!"
    play cards
play _ = return ()
